### PR TITLE
Add the generator of the docker-compose.yaml file

### LINKED ISSUE #118: Create Docker-Compose generator
This would be a react module that would essentially be a form that fills out a docker-compose template with user values and provides the .yml file for download.

### PR DESCRIPTION
**Pull request by extSunset**
Resolves #118 . It's still a bit raw. The design could be better, it might be worth generating the file on the first render. I need to solve the following warnings:
```
Warning: Can't perform a React state update on an unmounted component
``` 
```
Warning: State updates from the useState() and useReducer() Hooks don't support the second callback argument. To execute a side effect after rendering, declare it in the component body with useEffect()
```
This is my first time programming in React and JS, so the code may be bad. I'm hoping for some hints on how to improve this Pull Request, or even make changes themselves.

### CODE REVIEW

**Comment by EmilFattakhov** [comment]
Really love to see this issue being worked on, thanks a lot for picking it up @extSunset!
I like the overall approach. I'm inclining towards having the substrate-cli docker section untouched, but rather adding a separate page that helps users build their docket configuration. @ImmaZoni, what's your opinion on that?

I'll thoroughly review the code tomorrow and add a full review. Keep up the excellent work!

File: src/components/DockerFileGenerator/index.jsx [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    // Initial state for form data
+    const [formData, setFormData] = useState({
+        nodePort: '30333',
+        nodeDsnPort: '30433',
+        farmerPort: '30533',
+        nodeName: 'subspace',
+        snapshot: '',
+        nodeData: '',
+        farmerData: '',
+        rewardAddress: 'st6GBwATPqtBkK5y4uXbV52euszPpFPw7wmkF8FywEqJaf8uP',
```
Comment (EmilFattakhov) [inline src/components/DockerFileGenerator/index.jsx:17]: Let's maybe keep the default reward address empty. I'm afraid that some users might think that the "default" configuration suits them and actually forget to replace this `rewardAddress` with their own reward address.

File: docs/protocol/substrate-cli.mdx [hunk 2]
```diff
@@ -107,91 +108,13 @@ We will be downloading two files for your respective operating system.
...
-        # Replace `WALLET_ADDRESS` with your Polkadot.js wallet address
-        "--reward-address", "WALLET_ADDRESS",
-        # Replace `PLOT_SIZE` with plot size in gigabytes or terabytes, for example 100G or 2T (but leave at least 60G of disk space for node and some for OS)
-        "path=/var/subspace,size=PLOT_SIZE",
-      ]
-      volumes:
-        node-data:
-        farmer-data:
-  ```
+  1. Generate `docker-compose.yaml` file:
```
Comment (EmilFattakhov) [inline docs/protocol/substrate-cli.mdx:111]: Let's make instructions slightly more clear.
1. Input your parameters in the provided form and click `Generate`
2. Copy the generated `docker-compose.yml` file (the generated file should be in this section)

All latter instructions are clear and could be added next.

**Review by EmilFattakhov — CHANGES_REQUESTED** [review]
This looks AWESOME @extSunset, impressive work! This is so easy to use and I love the styling!

Left a few suggestions to make instructions more clear, but other than then, it's perfect!

**Comment by EmilFattakhov** [comment]
> Ports are checked for validity (1024-65535). All snapshots for Gemini-3f are obtained via GitHub API. Plot size only supports bytes, gigabytes and terabytes. Reward Address is also checked for validity. Node Data/Farmer Data can be either empty or absolute UNIX-like paths.

Love the way you tackle the checks via regex!

**Comment by EmilFattakhov** [comment]
> It might be worth applying this technique to documentation in general and getting the latest release for the current network (now Gemini-3f) via the GitHub API for both Pulsar and Substrate CLI.

Unfortunately, this might not work the best for us as versioning might become quite confusing if we enable auto-updated. 
But overall I think we should be moving toward automatization of the updates, as it becomes quite time-confusing. Appreciate your idea and implementation here!

**Comment by EmilFattakhov** [comment]
> My attempt to improve the generator interface. I had to resort to help, but I still think the interface could be better.

Looks beautiful IMO and you even added dark/light mode support.

**Comment by EmilFattakhov** [comment]
> It's probably better to create a custom select to make it looks native.

Custom select are a bit painful to work on. But generally speaking, I agree that custom selects look better in most cases. Not a deal breaker though.

**Comment by EmilFattakhov** [comment]
> I'd like to pack all commits into one when pull request will be ready to go live.

We can "squash and merge" when time comes, so don't worry about it :)

File: src/components/DockerFileGenerator/index.jsx (hunk 1 above)
Comment (EmilFattakhov) [reply src/components/DockerFileGenerator/index.jsx:17]: Yes, either of these options are good IMO.

File: docs/protocol/substrate-cli.mdx (hunk 2 above)
Comment (EmilFattakhov) [reply docs/protocol/substrate-cli.mdx:111]: Before the form, please :)

**Comment by EmilFattakhov** [comment]
> @EmilFattakhov I changed the interface a bit (tried to make it more harmonious and compact). Could you please give me your opinion?

I think it's a move in the right direction. It's more compact while being still very user-friendly. I like it! 
If that's not too complicated to try out, what do you think about making form a grid with two columns? If that's a stretch task, it's completely fine to ignore it.

**Comment by EmilFattakhov** [comment]
> I had some questions about the colour scheme. Why is --ifm-color-primary much lighter than --ifm-color-primary-lighter? It's supposed to be darker, isn't it? Or am I misunderstanding something?

That would have been my understanding too. Sorry, can't really elaborate much as I haven't touched the CSS file.

File: src/components/DockerFileGenerator/index.jsx (hunk 1 above)
Comment (EmilFattakhov) [reply src/components/DockerFileGenerator/index.jsx:17]: Alternatively, you can add this exact string at your REGEX validation and check if the address is valid and check if the address doesn't match the default one. Does my suggestion make sense? Happy to elaborate if needed.

**Comment by jim-counter** [comment]
I'm really pleased to read through the collaborative comments on this PR @extSunset. Looking forward to seeing a deploy preview so I can have a play. Thanks!

**Comment by EmilFattakhov** [comment]
> This is a bit off topic, but I noticed it while I was working on the generator. Does a light theme have to be that blinding? Some elements get lost in such brightness.

Agree, I also noticed that it could be a bit too bright at times. Let's not address this as a part of this PR though.

**Comment by EmilFattakhov** [comment]
> @EmilFattakhov I tried splitting the interface into 2 columns, worked out pretty well. Could you please give me your opinion?

I think it looks great and it definitely helped cut the page's length. Good job!

**Review by EmilFattakhov — CHANGES_REQUESTED** [review]
Once again, thank you very much @extSunset, you've put a lot of time and effort into this PR and it shows! I love the styling, splitting up into two columns is great and an error message by the default wallet address is super helpful! 

Two last pieces of feedback:

1. Let's put the `Docker Compose File Generator` into the first instruction step, right after
`Input your parameters in the provided form and click Generate. Don't forget to specify your reward address instead a default one!`

There is also a small typo in this sentence, "reward address instead a default one" -> "reward address instead of a default one"

2. Please remove `(the generated file should be in this section)` from the "Copy the generated docker-compose.yaml file (the generated file should be in this section)".


Once those are fixed, I believe PR is good to be merged!

**Review by EmilFattakhov — APPROVED** [review]
Thank you for implementing suggestions @extSunset, everything looks great.

Looking forward to your other contributions, keep it up!

**Pull request merged**